machinelearning/tensorflow.md

```diff
@@ -0,0 +1,415 @@
+
+```
+#clone the tf serving repository
+git clone https://github.com/tensorflow/serving.git
+cd serving
+
+# pull https://github.com/tensorflow/serving/pull/1953
+git fetch origin pull/1953/head:tfs_aarch64
+
+# pull https://github.com/tensorflow/serving/pull/1954
+git fetch origin pull/1954/head:tfs_docker_aarch64
+
+# merge them
+git checkout tfs_aarch64
+git merge tfs_docker_aarch64
+
+# and then invoke the docker build script to trigger mkl aarch64 config build
+docker build -f tensorflow_serving/tools/docker/Dockerfile.devel-mkl-aarch64 -t tfs:mkl_aarch64 .
+
+# command to launch serving api with oendnn+acl backend, and BF16 kernels. Here is an example for launching resnet graph
+docker run -p 8501:8501 --name tfserving_resnet --mount type=bind,source=/tmp/resnet,target=/models/resnet -e MODEL_NAME=resnet -e TF_ENABLE_ONEDNN_OPTS=1 -e DNNL_DEFAULT_FPMATH_MODE=BF16 -e -t tfs:mkl_aarch64
+
+```
+
+# Enable XLA optimizations
+While the Grappler optimizer covers majority of the networks, there are few scenarios where either the Grappler optimizer can't optimize the generic graph or the runtime kernel launch overhead is simply not acceptable. XLA addresses these gaps by providing an alternative mode of running models: it compiles the TensorFlow graph into a sequence of computation kernels generated specifically for the given model. Tenforflow-2.9.0 supports aarch64 xla backend with eigen runtime. In order to accelerate the xla ops with acl (arm compute library) runtime, one can cherrypick the [PR](https://github.com/tensorflow/tensorflow/pull/55534) and build tensorflow from sources.
+```
+#clone tensorflow repository
+git clone https://github.com/tensorflow/tensorflow.git
+cd tensorflow
+
+#pulll https://github.com/tensorflow/tensorflow/pull/55534
+git fetch origin pull/55534/head:xla_acl
+
+git checkout xla_acl
+
+# follow the steps from the building from sources section
+#set the build configuration
+export HOST_C_COMPILER=(which gcc)
+export HOST_CXX_COMPILER=(which g++)
+export PYTHON_BIN_PATH=(which python)
+export USE_DEFAULT_PYTHON_LIB_PATH=1
+export TF_ENABLE_XLA=1
+export TF_DOWNLOAD_CLANG=0
+export TF_SET_ANDROID_WORKSPACE=0
+export TF_NEED_MPI=0
+export TF_NEED_ROCM=0
+export TF_NEED_GCP=0
+export TF_NEED_S3=0
+export TF_NEED_OPENCL_SYCL=0
+export TF_NEED_CUDA=0
+export TF_NEED_HDFS=0
+export TF_NEED_OPENCL=0
+export TF_NEED_JEMALLOC=1
+export TF_NEED_VERBS=0
+export TF_NEED_AWS=0
+export TF_NEED_GDR=0
+export TF_NEED_OPENCL_SYCL=0
+export TF_NEED_COMPUTECPP=0
+export TF_NEED_KAFKA=0
+export TF_NEED_TENSORRT=0
+
+./configure
+
+#issue bazel build command with 'mkl_aarch64' config to enable onednn+acl backend
+bazel build --verbose_failure -s --config=mkl_aarch64  //tensorflow/tools/pip_package:build_pip_package //tensorflow:libtensorflow_cc.so //tensorflow:install_headers
+
+#Create and install the wheel
```

Comment:
not consistent with caps on the first word (or spaces)

Reply:
will fix.